clang/unittests/Analysis/FlowSensitive/TypeErasedDataflowAnalysisTest.cpp

```diff
@@ -975,6 +994,35 @@ TEST_F(WideningTest, DistinctValuesWithSamePropertiesAreEquivalent) {
       });
 }
 
+TEST_F(WideningTest, DistinctValuesWithDifferentPropertiesWidenedToTop) {
+  std::string Code = R"(
+    void target(bool Cond) {
+      int *Foo;
+      int i = 0;
+      Foo = nullptr;
+      while (Cond) {
+        Foo = &i;
+      }
+      (void)0;
+      /*[[p]]*/
+    }
+  )";
+  runDataflow(
+      Code,
+      [](const llvm::StringMap<DataflowAnalysisState<NoopLattice>> &Results,
+         ASTContext &ASTCtx) {
+        ASSERT_THAT(Results.keys(), UnorderedElementsAre("p"));
+        const Environment &Env = getEnvironmentAtAnnotation(Results, "p");
+
+        const ValueDecl *FooDecl = findValueDecl(ASTCtx, "Foo");
+        ASSERT_THAT(FooDecl, NotNull());
+
+        const auto *FooVal = Env.getValue(*FooDecl);
+        EXPECT_TRUE(areEquivalentValues(*FooVal->getProperty("is_null"),
```

Comment:
I think we also want to test that `getProperty()` returns non-null?

Easiest way might be `*cast<BoolValue>(FooVal->getProperty("is_null"))`. (`cast` asserts that its argument is non-null.)

Reply:
I prefer direct assertion, because it results in a cleaner error message on failure (crashing the test process vs exiting the test with a test-assertion failure).

Reply:
Looks good!